R/common-kmeans.R

```diff
@@ -0,0 +1,430 @@
+    ix <- lineMagnitude(px, py, x1, y1)
+    iy <- lineMagnitude(px, py, x2, y2)
+    if (ix > iy)  ans <- iy
+    else ans <- ix
+  } else {
+    ## Intersecting point is on the line, use the formula
+    ix <- x1 + u * (x2 - x1)
+    iy <- y1 + u * (y2 - y1)
+    ans <- lineMagnitude(px, py, ix, iy)
+  }
+  return(ans)
+}
+
+#' @title
+#' Compute the distance of a point from a line 
+#'
+#' @description compute the distance of a point from a line given the 
+#' x, y axes of the point and the slope and intercept of the line.
+#' @param x A numeric vector, x axis of the points
+#' @param y A numeric vector, y axis of the points
+#' @param slope A number, the slope of the line
+#' @param intercept A number, the intercept of the line 
+#' @return A numeric vector, the length of the pointS from the line. 
+#'
+#' @export
+#' @references \url{http://healthcare.ai}
+#' @references This helper funtion is originally found here 
+#' \url{http://paulbourke.net/geometry/pointlineplane/pointline.r}
+#' @references \url{https://github.com/bryanhanson/ChemoSpecMarkeR/blob/master/R/findElbow.R}
+#' @seealso \code{\link{healthcareai}}
+#' @examples
+#' distancePointLine(c(2,0),c(3,2),-0.5,2.5)
+distancePointLine <- function(x, y, slope, intercept) {
+  x1 <- x - 10
+  x2 <- x + 10
+  y1 <- x1*slope + intercept
+  y2 <- x2*slope + intercept
+  ans <- distancePointSegment(x,y, x1,y1, x2,y2)
+  return(ans)
+}
+
+#' @title
+#' Find the elbow in a curve
+#'
+#' @description finds the elbow of a curve that is concave to the line connecting
+#' the first and last points. 
+#' @param y A numeric vector of more than 2 elements  w/o NA's
+#' @return A number, the index of the elbow point. 
+#'
+#' @export
+#' @references \url{http://healthcare.ai}
+#' @references This function is developed based on the following function in 
+#' package ChemoSpecMarkeR
+#' \url{https://github.com/bryanhanson/ChemoSpecMarkeR/blob/master/R/findElbow.R}
+#' @references The idea behind the function can be found here
+#' \url{https://stackoverflow.com/questions/2018178/finding-the-best-trade-off-point-on-a-curve/2022348#2022348}
+#' @seealso \code{\link{healthcareai}}
+#' @examples
+#' y <- c(8.5, 4.9, 2.8, 2.5, 1.9, 1.1, 1.1, 0.9)
+#' plot(y) # concave
+#' findElbow(y)
+#' 
+#' y <- c(6,5.5,4,2,1.5)
+#' plot(y) # not concave
+#' ## findElbow(y) will gave an error
+findElbow <- function(y) {
+  if (length(y) <= 2) 
+    stop("The curve has less than 2 points. At least three points in a curve are needed to find its elbow.")
+  # Add an index to argument values for easy plotting
+  DF <- data.frame(x = 1:length(y), y = y)
+  fit <- lm(y ~ x, DF[c(1,nrow(DF)),]) # 2 point 'fit'
+  m <- coef(fit)[2]
+  b <- coef(fit)[1]
+
+  # Check to make sure the data is concave
+  concave <- FALSE
+  use <- 2:(nrow(DF) - 1)
+  refpts <- m*DF$x[use] + b
+  if (all(refpts > DF$y[use]) | all(refpts < DF$y[use])) concave <- TRUE
+  if (!concave) stop("Your curve doesn't appear to be concave")
```

Comment:
Let's use `isTRUE` for these boolean checks

https://stat.ethz.ch/R-manual/R-devel/library/base/html/Logic.html